datafusion/physical-expr/src/simplifier/unwrap_cast.rs

```diff
@@ -0,0 +1,621 @@
+        Operator::GtEq => Operator::LtEq,
+        // Symmetric operators remain the same
+        Operator::Eq | Operator::NotEq => op,
+        // Other operators are not swappable
+        _ => op,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::expressions::{col, lit};
+    use arrow::datatypes::{DataType, Field, Schema};
+    use datafusion_common::ScalarValue;
+    use datafusion_expr::Operator;
+
+    /// Check if an expression is a cast expression
+    fn is_cast_expr(expr: &Arc<dyn PhysicalExpr>) -> bool {
+        expr.as_any().downcast_ref::<CastExpr>().is_some()
+            || expr.as_any().downcast_ref::<TryCastExpr>().is_some()
+    }
+
+    /// Check if a binary expression is suitable for cast unwrapping
+    fn is_binary_expr_with_cast_and_literal(binary: &BinaryExpr) -> bool {
+        // Check if left is cast and right is literal
+        let left_cast_right_literal = is_cast_expr(binary.left())
+            && binary.right().as_any().downcast_ref::<Literal>().is_some();
+
+        // Check if left is literal and right is cast
+        let left_literal_right_cast =
+            binary.left().as_any().downcast_ref::<Literal>().is_some()
+                && is_cast_expr(binary.right());
+
+        left_cast_right_literal || left_literal_right_cast
+    }
+
+    fn test_schema() -> Schema {
+        Schema::new(vec![
+            Field::new("c1", DataType::Int32, false),
+            Field::new("c2", DataType::Int64, false),
+            Field::new("c3", DataType::Utf8, false),
+        ])
+    }
+
+    #[test]
+    fn test_unwrap_cast_in_binary_comparison() {
+        let schema = test_schema();
+
+        // Create: cast(c1 as INT64) > INT64(10)
+        let column_expr = col("c1", &schema).unwrap();
+        let cast_expr = Arc::new(CastExpr::new(column_expr, DataType::Int64, None));
+        let literal_expr = lit(ScalarValue::Int64(Some(10)));
+        let binary_expr =
+            Arc::new(BinaryExpr::new(cast_expr, Operator::Gt, literal_expr));
+
+        // Apply unwrap cast optimization
+        let result = unwrap_cast_in_comparison(binary_expr, &schema).unwrap();
+
+        // Should be transformed
+        assert!(result.transformed);
+
+        // The result should be: c1 > INT32(10)
+        let optimized = result.data;
+        let optimized_binary = optimized.as_any().downcast_ref::<BinaryExpr>().unwrap();
+
+        // Check that left side is no longer a cast
+        assert!(!is_cast_expr(optimized_binary.left()));
+
+        // Check that right side is a literal with the correct type and value
+        let right_literal = optimized_binary
+            .right()
+            .as_any()
+            .downcast_ref::<Literal>()
+            .unwrap();
+        assert_eq!(right_literal.value(), &ScalarValue::Int32(Some(10)));
+    }
+
+    #[test]
+    fn test_unwrap_cast_with_literal_on_left() {
+        let schema = test_schema();
+
+        // Create: INT64(10) < cast(c1 as INT64)
```

Comment:
Can you also please test something where the cast doesn't work? Something with a column that is `uint8` for example, and the literal can not fit into the column: `Int64(-5) < CAST(c4 as Int64)

Though I realize in this case the predicate would always be true then 🤔